File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/LocationService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.Location;
import at.ac.tuwien.sepm.groupphase.backend.exception.ConflictException;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;

import java.util.List;

public interface LocationService {

    /**
     * Retrieve a list of all locations in the persistent data store
     *
     * @return list of all locations
     */
    List<Location> getAllLocations();

    /**
     * Get location specified by id.
     *
     * @param id id of the location to be retrieved
     * @return location that has the given id
     * @throws NotFoundException s thrown if no location with given id exists in the database
     */
    Location getLocationById(long id) throws NotFoundException;

    /**
     * Store new location in persistent data store
     *
     * @param location new location that should be persistet
     * @return returns entity that has been stored in the database
     * @throws ConflictException if this Location already exists in the database
     */
    Location createLocation(Location location) throws ConflictException;

    /**
     * Gets all locations which fullfill all the search parameters.
     *
     * @param name    of the location
     * @param street  of the location
     * @param city    of the location
     * @param country of the location
     * @param zipCode of the location
     * @return list of locations that fit the ssearch terms
     */
    List<Location> search(String name, String street, String city, String country, String zipCode);
}

File: frontend/src/app/services/location.service.ts
import {HttpClient, HttpParams} from '@angular/common/http';
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {Location} from '../dtos/location';
import {Globals} from '../global/globals';

@Injectable({
  providedIn: 'root'
})
export class LocationService {

  private locationBaseUri: string = this.globals.backendUri + '/locations';

  constructor(
    private httpClient: HttpClient,
    private globals: Globals
  ) {
  }


  /**
   * Get all the locations stored in the database.
   *
   */
  getAll(): Observable<Location[]> {
    return this.httpClient.get<Location[]>(this.locationBaseUri);
  }

  /**
   * Get location by id
   *
   * @param id id of the location to be retrieved
   * @returns retrieved location
   */
  getById(id: number): Observable<Location> {
    return this.httpClient.get<Location>(this.locationBaseUri + `/${id}`);
  }

  /**
   * Create location in backend
   *
   * @param location location to be stored in the backend
   * @returns persisted location
   */
  createLocation(location: Location): Observable<Location> {
    return this.httpClient.post<Location>(this.locationBaseUri, location);
  }

  search(name: string,
         street: string,
         city: string,
         country: string,
         zipCode: string): Observable<Location[]> {
    let params = new HttpParams();
    if (name !== undefined && name.trim() !== '') {
      params = params.set('name', name);
    }
    if (street !== undefined && street.trim() !== '') {
      params = params.set('street', street);
    }
    if (city !== undefined && city.trim() !== '') {
      params = params.set('city', city);
    }
    if (country !== undefined && country.trim() !== '') {
      params = params.set('country', country);
    }
    if (zipCode !== undefined) {
      params = params.set('zipCode', zipCode);
    }
    return this.httpClient.get<Location[]>(this.locationBaseUri + '/search', {params});
  }
}
